### DISCUSSION TITLE
`nix develop` shell incompatible with Cabal 3.12.1.0?

### DISCUSSION THREAD

**Discussion by peterbecich** [srid/haskell-flake#339] [Q&A] [answered]
`cabal` and `cabal-syntax` have been recently updated to 3.12.1.0 on NixPkgs.

`cabal` and `cabal-syntax` build successfully themselves:
```
nix build nixpkgs/haskell-updates#haskellPackages.Cabal_3_12_1_0
nix build nixpkgs/haskell-updates#haskellPackages.Cabal-syntax_3_12_1_0
```

Would you help me with this error?
On
https://github.com/peterbecich/hackage-server/tree/nix-cabal-3.12

`nix develop` yields an error building `cabal-install-3.12.1.0`
```
src/Distribution/Client/Glob.hs:24:1: error:
    Could not find module ‘Distribution.Simple.Glob.Internal’
    Perhaps you meant
      Distribution.Simple.GHC.Internal
      Distribution.Simple.GHC.Internal
      Distribution.Simple.Program.Internal (needs flag -package-id Cabal-3.10.3.0)
    Use -v (or `:set -v` in ghci) to see a list of the files searched for.
   |
24 | import Distribution.Simple.Glob.Internal
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

and

```
src/Distribution/Client/Errors.hs:861:21: error: [GHC-76037]
    Not in scope: type constructor or class ‘VerboseException’
    |
861 | instance Exception (VerboseException CabalInstallException) where
    |                     ^^^^^^^^^^^^^^^^

src/Distribution/Client/Errors.hs:862:23: error: [GHC-76037]
    Not in scope: type constructor or class ‘VerboseException’
    |
862 |   displayException :: VerboseException CabalInstallException -> [Char]
    |                       ^^^^^^^^^^^^^^^^

src/Distribution/Client/Errors.hs:863:21: error: [GHC-76037]
    Not in scope: data constructor ‘VerboseException’
    |
863 |   displayException (VerboseException stack timestamp verb cabalInstallException) =
    |                     ^^^^^^^^^^^^^^^^
```

**Comment by srid** [comment]
Just a quick suggestion, since I recall facing similar issues.

https://github.com/peterbecich/hackage-server/commit/c0de63ce061a44109ca4049e199b205ae83eba80#diff-206b9ce276ab5971a2489d75eb1b12999d4bf3843b7988cbe8d687cfde61dea0R51-R54

It probably is not a good idea to override cabal packages in the package set (i.e., `settings`). If you just want to use the newer cabal, maybe best to specify them directly in the devShell (or haskell-flake's `devShell.tools`).

**Reply by peterbecich** [reply]
These are required for `nix build` to succeed; without them the error is
```
       > Error: Setup: Encountered missing or private dependencies:
       > Cabal >=3.12.1.0 && <3.14, Cabal-syntax >=3.12.1.0 && <3.14
```

**Reply by srid** [reply]
So [hackage-server uses the new cabal as dependency](https://github.com/haskell/hackage-server/pull/1322/files#diff-80d8280eac7e7395c335fe96148a837a1e7bcf1f8e3511fb84465551ed80cf47L151-R153).

In that case, yea, you do want the new cabal in the package set.

---

> `nix develop` yields an error building `cabal-install-3.12.1.0`

So, try overriding [`cabal-install`](https://hackage.haskell.org/package/cabal-install) as well, in `settings`?

**Reply by peterbecich** [reply]
Added
```
cabal-install = { super, ... }: { custom = _: super.cabal-install_3_12_1_0; };
```
https://github.com/haskell/hackage-server/blob/adfcbd90eec840f222eb9cc54cfc84567d89a91a/flake.nix#L51-L58

`nix develop` yields same error with `VerboseException`
`nix build` succeeds as it did before

**Reply by srid** [reply]
@locallycompact - any ideas?

(As an aside: haskell-flake [also supports horizon package sets](https://github.com/srid/haskell-template/pull/123)).

**Reply by peterbecich** [reply]
Thanks @srid, here is the error reproduced in Github Actions:
https://github.com/haskell/hackage-server/actions/runs/10023429550/job/27704328133?pr=1323
https://github.com/haskell/hackage-server/pull/1323

**Reply by peterbecich** [reply]
~Maybe this will make a difference: https://github.com/NixOS/nixpkgs/pull/328757~ it does not

**Reply by srid** [reply, marked as answer]
As far as I can see, the fix for this nevertheless can only happen on upstream, since haskell-flake simply wraps whatever `nixpkgs` provides.  `settings` & `packages` are transformed into an overlay and then added to basePackages using `extend`:

https://github.com/srid/haskell-flake/blob/cd449f1c04175efdf5b553302d22916640090066/nix/modules/project/outputs.nix#L86-L91

**Reply by peterbecich** [reply]
Very good, I'll try to define a derivation that reproduces the issue minimally and bring it up there, thanks again @srid

**Reply by peterbecich** [reply]
Bingo, this reproduces it
```
nix build nixpkgs/haskell-updates#haskellPackages.cabal-install 
```
great

filed issue https://github.com/NixOS/nixpkgs/issues/328765

**Reply by locallycompact** [reply]
Yeah we get this occasionally but horizon will catch it in CI. nixpkgs doesn't seem to build haskellPackages before merging PRs which leads to this.
